recipes/newrelic/apm/java/linux.yml

```diff
@@ -30,71 +32,37 @@ logMatch:
-    if  [[ -z $INIT_SYSTEM || "$INIT_SYSTEM" != "systemd" ]]; then
-      echo -e "Unsupported process management system: $INIT_SYSTEM" >&2
+    cutInstalled=$(which cut 2> /dev/null)
+    if [[ -z "$cutInstalled" ]]; then
+      echo -e "This installation recipe for the New Relic Java Agent on Linux requires 'cut' to be installed." >&2
       exit 132
     fi
 
-    TMP_DIR=$(mktemp -d /tmp/newrelic.XXXXXX)
-    cd $TMP_DIR
-
-    isYumInstalled=$(which yum 2>&1 || true)
-    if [[ -n "$isYumInstalled" ]]; then
-      curl -s -O https://download.newrelic.com/install/java/introspector/latest/nri-introspector-java-0.1.0_SNAPSHOT-2.x86_64.rpm
-      yum install -y -q nri-introspector-java-0.1.0_SNAPSHOT-2.x86_64.rpm 2> /dev/null || true
-    else
-      curl -s -O https://download.newrelic.com/install/java/introspector/latest/nri-introspector-java_0.1.0_SNAPSHOT-2_amd64.deb
-      apt-get install -y -qq ./nri-introspector-java_0.1.0_SNAPSHOT-2_amd64.deb 2> /dev/null || true
-    fi
-
-    # check if any tomcat instances are service-managed
-    nri-lsi-java -list | tr -d '[]' | tr ',' '\n' > $TMP_DIR/all_java_processes
-    for JAVA_PID in $(cat $TMP_DIR/all_java_processes)
-    do
-      INTROSPECTION_DATA=$(nri-lsi-java -introspect ${JAVA_PID})
-      IS_JAVA_TOMCAT=$(echo $INTROSPECTION_DATA | grep "org.apache.catalina.startup.Bootstrap" | grep -v grep | wc -l)
-      if [ "$IS_JAVA_TOMCAT" -gt 0 ]; then
-        IS_MANAGED_TOMCAT=$(systemctl status $JAVA_PID | head -n1 | grep -oP "(?<=. )\K.+(?=\.service)" | wc -l)
-        if [ "$IS_MANAGED_TOMCAT" -gt 0 ]; then
-          rm -rf $TMP_DIR
-          exit 0
-        fi
-      else
-        touch $TMP_DIR/non_tomcat_java_process
+    # Map of tool names to the associated error code
+    required_tools_and_error_codes="grep:132 cat:132 tee:132"
+    for tuple in $required_tools_and_error_codes; do
+      tool=$(echo ${tuple} |cut -d':' -f1)
+      code=$(echo ${tuple} |cut -d':' -f2)
+      toolInstalled=$(which ${tool} | wc -l)
+      if [[ "$toolInstalled" -eq 0 ]]; then
+        echo -e "This installation recipe for the New Relic Java Agent on Linux requires '${tool}' to be installed." >&2
+        exit ${code}
       fi
     done
 
-    if [[ -f $TMP_DIR/non_tomcat_java_process ]]; then
-      echo -e "Java processes other than Tomcat have been detected" >&2
-    else
-      echo -e "Tomcat processes were detected, but no running java processes found" >$2
-    fi
-
-    if [[ -n "$isYumInstalled" ]]; then
-      yum remove -y -q nri-introspector-java 2> /dev/null
-    else
-      apt remove -y -qq nri-introspector-java 2> /dev/null
-    fi
-
-    rm -rf $TMP_DIR
-    exit 132
+    exit 0
 
-validationNrql: "SELECT count(*) FROM ApplicationAgentContext WHERE host LIKE '{{.HOSTNAME}}%' SINCE 10 minutes AGO"
+validationNrql: "SELECT count(*) FROM ApplicationAgentContext where agent.language = 'java' and (host like '{{.HOSTNAME}}%' or container.id is not null) SINCE 10 minutes AGO"
```

Comment:
@mbruzina I don't know if the container.is is appropriate without checking the hostname, because the validation would succeed for a subsequent on a different host with a container even if somehow the instrumentation fails. Maybe just something to be aware in the future, maybe not something to address right now.

Reply:
Yes, this is brittle at the moment.  I'm not considering a showstopper because as written it permits both containerized and non-containerized installations to succeed with a fair amount of certainty that the java agent is installed.   

The dilemma is hostname changes to the containerId when installing/dynamically attaching within a container.  If there's a way to somehow dynamically change the validation query, we can, but I'm not aware of one ATM.  I'd like to use appName and agent language if possible.

Reply:
At some point, we should switch to use tracer event, not validating against a specifc app. That's what we do for some of the other apm installatations.